<a href="https://colab.research.google.com/github/shakedseller/Module-1-Class-1-FQ-Shaked-Seller/blob/main/MCON232_FQ_Homework_1_Lists_Strings_FileIO.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# MCON 232 FQ • Homework 1 • Lists (1D) + Strings + File I/O (Colab-friendly)

**Estimated time:** ~75 minutes  
**Due:** _(your date here)_  
**Name:** ___________________________

### NOTE :  
- although I pace the homework, this is guidance and aspirational, it is okay to take significantly longer.

### Rules (important)
- **list comprehensions is NOT required** for this homework.  
- **slicing is NOT required** (like `lst[1:4]`) unless a problem explicitly tells you to.  
- Use **loops** and **basic list methods** (`append`, `insert`, `pop`, `remove`).  
- For string cleanup, use: `strip()` and `split()`.
- Keep your code readable: meaningful variable names, blank lines, and short comments.  

> Tip: Run the notebook **top to bottom**. Some later parts rely on earlier cells.


In [ ]:
# Optional: quick check that you're running in a notebook environment
print("Notebook is running ✅")


## 🌱 Support   
Note : please return to the slides and to Gaddis text for support.  Remember to post in Discussions if there is a significant concern.  Thank you.

### 🧺 Lists (1D)
Remember:
- A list is just a collection of items in order.
- You can:
  - find its length with `len(my_list)`
  - access items using indexes: `my_list[0]`, `my_list[-1]`
  - change items by assignment: `my_list[2] = "new value"`
- Common list tools:
  - `append(x)` → add to the end
  - `insert(i, x)` → add at a position
  - `pop()` → remove and return an item
  - `remove(x)` → remove a specific value

When in doubt: **print the list** and look at it.

---

### 🧵 Strings
Strings are just text — and text often needs cleaning.

Two very helpful tools:
- `strip()` removes extra whitespace (including `\n`)
- `split(",")` breaks a string into parts

Example:
```python
line = "Alice,90\n"
clean = line.strip()
parts = clean.split(",")
```
---

### 📂 Files : A Gentle Reminder About `open()`

When working with files, everything starts with `open()`.

In this notebook:
- We **open a file in write mode first** to create it
- Then we **open the same file in read mode** to read it back

👉 **Please run the file-writing cell first.**  
If you skip it, the file will not exist yet.

Remember:
- `"w"` = write (create the file)
- `"r"` = read
- Each line you read is a **string**
- Clean it with `strip()`
- Break it apart with `split(",")`
- Convert numbers with `int()`

Take it slowly.  
If you’re unsure, print intermediate values.  
This is exactly how file work looks in real programs. 💙

---

### 🔁 Loops

Loops help you think step-by-step.

**Common patterns:**
- **Counting** (start at 0, add 1 when something matches)
- **Accumulating** (sum, product)
- **Building a new list** (start empty, append carefully)
- **Tracking positions** (use `range(len(my_list))`)

There is no rush. **One loop at a time is perfect.**


## Part 1 — Refresher: Lists & Strings (10 minutes)

These are quick warmups to get your hands moving:
- `len(list)`
- indexing (first / last)
- updating list elements
- basic list methods
- string cleanup with `strip()` and `split()`


### Task 1A — Fix the list (easy warmup)

You will practice:
- length
- first / last element
- replacing an element


In [2]:
animals = ["cat", "dog", "fish", "bird"]

# TODO:
# 1) Print the length of the list
# 2) Print the first item
# 3) Print the last item (no slicing)
# 4) Replace "fish" with "hamster"
# 5) Print the updated list

# Write your code below:

print(f"Here is the length of the list: {len(animals)}")
print(f"The first item in the list: {animals[0]}")
print(f"The last item in the list: {animals[len(animals)-1]}")
animals[2] = "hamster"
print(f"The updated list:\n{animals}")


Here is the length of the list: 4
The first item in the list: cat
The last item in the list: bird
The updated list:
['cat', 'dog', 'hamster', 'bird']


### Task 1B — List methods practice

You **must** use the method named in each step.


In [8]:
nums = [10, 20, 30]

# TODO:
# 1) append 40
# 2) insert 99 at index 1
# 3) pop the last value into a variable called last_val
# 4) remove 20
# 5) print nums and last_val

# Write your code below:
nums.append(40)
nums.insert(1, 99)
last_val = nums.pop(len(nums)-1)
nums.remove(20)
print(f"List nums: {nums}")
print(f"Variable last_val: {last_val}")


List nums: [10, 99, 30]
Variable last_val: 40


### Task 1C — String refresher: `strip()` and `split()`

A file line often looks like: `"Alice,90\n"`  
We want to clean it and break it into parts.

Example:
- `line.strip()` removes whitespace/newlines at both ends
- `clean.split(",")` splits into a list of strings


In [9]:
line = "Alice,90\n"

# TODO:
# 1) Create a variable clean_line that removes the newline using strip()
# 2) Split clean_line by comma into a list called parts
# 3) Print clean_line and parts

# Write your code below:

clean_line = line.strip("\n")
parts = clean_line.split(",")
print(f"clean_line: {clean_line}")
print(f"parts: {parts}")


clean_line: Alice,90
parts: ['Alice', '90']


## Part 2 — Thinking with Loops (20 minutes)

These problems are about **patterns**:
- counting
- filtering into a new list
- summing and multiplying with accumulators
- tracking an index with a loop


### Task 2A — Count & filter (no comprehensions)

Given a list of numbers:
1) Count how many are **>= 10**
2) Build a new list containing only numbers **>= 10**


In [10]:
nums = [3, 8, 2, 11, 5, 10, 10, 1]

# Required:
# count_big  -> how many numbers >= 10
# big_nums   -> list of numbers >= 10

count_big = 0
big_nums = []

# TODO: write your loop here
for i in range(len(nums)):
  if nums[i] >= 10:
    big_nums.append(nums[i])
    count_big += 1

print("count_big:", count_big)
print("big_nums:", big_nums)


count_big: 3
big_nums: [11, 10, 10]


### Task 2B — Two accumulators

Compute:
1) `total_sum` of the list
2) `even_product` = product of ONLY the even numbers  
   - Start `even_product` at 1  
   - Multiply only when the number is even


In [12]:
nums = [4, 7, 2, 9, 10, 3, 8]

total_sum = 0
even_product = 1

# TODO: write your loop here
for i in range(len(nums)):
  total_sum += nums[i]
  if nums[i] % 2 == 0:
    even_product *= nums[i]

print("total_sum:", total_sum)
print("even_product:", even_product)


total_sum: 43
even_product: 640


### Task 2C — “Where is it?” (index tracking)

Find the index of the **first** `"hello"` using a loop.  
Do **not** use `.index()` for this one.
*** HINT *** this is a perfect time to try out enumerate

In [15]:
words = ["hi", "hello", "hey", "hello", "hola"]

first_hello_index = -1  # keep -1 if not found

# TODO: write your loop here
# Hint: use range(len(words)) so you can check indexes.
for index, word in enumerate(words):
  if word == "hello":
    first_hello_index = index
    break

print("first_hello_index:", first_hello_index)


first_hello_index: 1


## Part 3 — File I/O Refresher (Colab-Friendly) (25 minutes)

In Google Colab, your notebook runs on a temporary computer.  
So we will **write a file first**, then **read it back**.

We will work with a file named: `grades.txt`

Each line contains:
```
Name,Score
```
Example:
```
Alice,90
```


### Step 3A — Create the input file (provided)

Run this cell **as-is** (do not change it).


In [16]:
# DO NOT CHANGE THIS CELL
sample_text = """Alice,90
Ben,72
Cara,88
Dan,95
Eva,72
"""

with open("grades.txt", "w") as f:
    f.write(sample_text)

print("grades.txt created ✅")


grades.txt created ✅


### Step 3B — Read and parse the file into lists

Build two **parallel lists**:
- `names`  (strings)
- `scores` (ints)

Requirements:
- Use a loop to read each line
- Use `strip()` to remove the newline
- Use `split(",")` to separate name and score
- Convert score to `int`

***NOTE*** if you are not certain how the string methods work, please look up on the internet.


In [18]:
names = []
scores = []

# TODO:
# 1) Open grades.txt for reading
# 2) Loop through each line
# 3) Strip newline
# 4) Split by comma into name and score
# 5) Append name to names
# 6) Convert score to int and append to scores

# Write your code below:
gradefile = open("grades.txt", "r")

line = gradefile.readline()
while line != "":
  cleaned = line.strip("\n")
  splitline = cleaned.split(",")
  name = splitline[0]
  score = splitline[1]

  names.append(name)
  scores.append(int(score))

  line = gradefile.readline()


print("names:", names)
print("scores:", scores)


names: ['Alice', 'Ben', 'Cara', 'Dan', 'Eva']
scores: [90, 72, 88, 95, 72]


### Step 3C — Analyze the data (thinking)

Using `names` and `scores`, compute:

1) `avg_score` (average score as a float)
2) `highest_score` and the `highest_name` who earned it
3) `count_72` = how many students earned exactly 72
4) `passing_names` = list of students with score **>= 75**

**Do not use list comprehensions.**


In [19]:
# Required variables:
avg_score = 0.0
highest_score = -1
highest_name = ""
count_72 = 0
passing_names = []

# TODO:
# - compute total
# - compute avg_score
# - find highest score + name
# - count 72s
# - build passing_names

# Note from Shaked: I put them all in separate loops because it helps me stay more organized, but I can try to condense it into one loop if you prefer.
total = 0
for i in range(len(scores)):
  total += scores[i]

avg_score = total / len(scores)

for i in range(len(scores)):
  if scores[i] > highest_score:
    highest_score = scores[i]
    highest_name = names[i]

for i in range(len(scores)):
  if scores[i] == 72:
    count_72 += 1

for i in range(len(scores)):
  if scores[i] >= 75:
    passing_names.append(names[i])

print("avg_score:", avg_score)
print("highest:", highest_name, highest_score)
print("count_72:", count_72)
print("passing_names:", passing_names)


avg_score: 83.4
highest: Dan 95
count_72: 2
passing_names: ['Alice', 'Cara', 'Dan']


### Step 3D — Write a report file

Create a file named `report.txt` that contains lines like:

- `Average: 83.4`
- `Top Student: Dan (95)`
- `Count of 72: 2`
- `Passing: Alice, Cara, Dan, Eva`

Then read and print the contents of `report.txt` to prove it worked.


In [22]:
# TODO:
# 1) Open report.txt in write mode and write the 4 lines
# 2) Then open report.txt in read mode and print it

# Write your code below:
reportfile = open("report.txt", "w")

reportfile.write(f"Average: {avg_score}\n")
reportfile.write(f"Top Student: {highest_name} ({highest_score})\n")
reportfile.write(f"Count of 72: {count_72}\n")
reportfile.write(f"Passing: {passing_names}\n")

reportfile.close()

reportfile = open("report.txt", "r")
read_reportfile = reportfile.read()
print(read_reportfile)


Average: 83.4
Top Student: Dan (95)
Count of 72: 2
Passing: ['Alice', 'Cara', 'Dan']



## Part 4 — Remove Duplicates (20 minutes)

You are not using sets yet, so we’ll solve this with a loop.

**Goal:** Build `unique_nums` containing each number only once, in the order it first appears.

Example:
- Input:  `[2, 5, 2, 9, 5, 1]`
- Output: `[2, 5, 9, 1]`

Hint:  
Use this pattern:
- for each `n` in `nums`  
  - if `n` is **not** already in `unique_nums`, append it


In [25]:
nums = [2, 5, 2, 9, 5, 1, 9, 9, 3]

unique_nums = []

# TODO: write your loop here
for n in nums:
  if n not in unique_nums:
    unique_nums.append(n)

print("unique_nums:", unique_nums)


unique_nums: [2, 5, 9, 1, 3]


### Optional (bonus) — How many duplicates were removed?

Compute `duplicates_removed`.

Hint:
- original count is `len(nums)`
- unique count is `len(unique_nums)`


In [27]:
# BONUS:
duplicates_removed = 0

# TODO:
# duplicates_removed = len(nums) - len(unique_nums)
duplicates_removed = len(nums) - len(unique_nums)

print("duplicates_removed:", duplicates_removed)


duplicates_removed: 4


## Submit

When finished, **download your notebook** and submit it according to your class instructions.

✅ You should have completed:
- basic list operations
- loop patterns (count, filter, accumulate, track index)
- string cleanup with `strip()` and parsing with `split()`
- file I/O (write → read → analyze → write report)
- a simple algorithmic thinking problem (remove duplicates)


## Please answer these questions  (double click to answer)
#### How long did this homework take you, excluding break time ?



#### How did you feel about this homework : it was easy, it was okay, it was very challenging.  And why ?

This homework was okay. It was a good refresher for file I/O, strings, loops, and lists. It took me about 35 minutes. Looking forward to seeing you on Sunday Bez"H, wishing you a peaceful and blessed Shabbat!!